### ISSUE TITLE
Support multiple Consul Property Sources locations to pull and bind properties into application beans

### ISSUE THREAD

**Issue by milind-git** [spring-cloud/spring-cloud-consul#303]
I am storing properties/configuration in consul kv store and then using spring cloud consul to bind property values into application beans and update them whenever there is an update to KV Store. I am using spring.cloud.consul.config.format=FILES .
My Property files are located as :
1> App specific property files  myapp.properties 
2> global property files are  global.properties ( this has high precedence than local property file if property key is same)

The way this would go inside KV store cleanly is :
```
/config/applications/<appname>/myapp.properties 
/config/global/global.properties
/config/xyz..... ( any other)
```

In fact ,there should  be a provision for 3rd tree structure considering the future extensibility.

with following bootstrap.yml , I can pull properties from myapp.properties from consul KV store dynamically  and bind them to application beans.

```
spring:
  application:
    name: myapp
  cloud:
    consul:
      config:
        enabled: true
        prefix: /config/applications/myapp
        format: FILES
```
However, the issue is , how do I pull properties from global.properties which is in  a different directory structure? prefix support a single value and doesn't support multiple comma separated values. Putting all properties files under one directory doesn't make sense as one is app-specific while other is global, when we are consul as store for our application configurations , we would like to be as organized as possible.

Using Spring profile doesn't suit as that has other uses ( like for example using it for loading environment specific files.

Can we enhance spring cloud consul to support this possibly by adding support for comma separated prefix ( with file names as well possibly) or some other alternative approach? Something similar to spring.config.name & spring.cloud.bootstrap.name (which can both handle comma separated lists of names).The requirement is to support storing multiple properties/configuration files with different locations and pull the values from such files into application beans ( considering the precedence of files for same property keys). This could be for properties, YAML files etc.

Any help is appreciated.

Thanks.

**Comment by spencergibb** [comment]
The default that applies to all applications is `application.properties`. You could also set `spring.cloud.consul.config.default-context=global` in bootstrap.properties.

**Comment by milind-git** [comment]
Thanks @spencergibb for prompt reply.
If I use spring.cloud.consul.config.default-context=global ( with prefix=config and applicationname=app1) , I would get following property sources generated :
/config/app1.properties, /config/app1.yaml , /config/app1.yml
and /config/global.properties, /config/global.yaml , /config/global.yml

I have several such application ( app2...appN) which means root config directory would have all app1.. to appN.properties file which is not organized.
The nice way to organize these is having a /config/apps/<appname>/app.properties 
for example
 config
            /apps
                      /app1
                               /app1.properties , 
                     /app2
                                /app2.properties 
and so on .

Similarly for shared files ,the location would be :
/config/shared/global.properties

So having a comma separated list for prefix ( with descending order of precedence) should be good enough to get this working.

If we can get this in , it means we can use spring cloud consul for storing properties/configuration in consul KV store and updating it dynamically which is a very strong feature in itself.

**Comment by jwedel** [comment]
We have a mixed set of applications, some spring-based some not. Both share Consul as configuration server.

As a spring application, we need to access some common configuration that is used for all services, e.g.:

    non-spring-app/key1
    spring-app/key2
    message-queue/key3

So from the point of view of `spring-app`, I need to access both `spring-app/key2` as well as `message-queue/key3`.

After examining the spring cloud consul sources, I came up with a configuration that is a bit hacky, but works:

`bootstrap.yml`:

```
spring:
  application.name: spring-app
  cloud:
    consul:
      config:
        prefix:
        defaultContext: /
        acl-token: ${CONSUL_ACL_TOKEN}
      enabled: true
```
*I left some parts of the configuration for brevity*

So, `prefix` is empty, `defaultContext` must not be empty since it is validated so I set it to `/` which works.

Now I can set e.g. `spring-app/server.port` and spring will pick it up. Additionally, I can directly access MQ properties in `application.yml`:

```
spring.rabbitmq:
    port: ${message-queue.port}
    username: ${message-queue.user}
    password: ${message-queue.password}
    host: ${message-queue.host}
```

However, I'm pretty sure that this was not the intended way to use the cloud consul configuration. I just wanted to add a scenario to the discussion.
